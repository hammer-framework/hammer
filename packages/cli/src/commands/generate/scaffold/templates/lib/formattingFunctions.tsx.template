import React from 'react'

import humanize from 'humanize-string'

const MAX_STRING_LENGTH = 150

export const formatEnum = (values: string | string[] | null | undefined) => {
  if (values) {
    if (Array.isArray(values)) {
      const humanizedValues = values.map((value) => humanize(value))
      return humanizedValues.join(', ')
    } else {
      return humanize(values as string)
    }
  }

  return ''
}

export const truncate = (value: string | number) => {
  const output = value?.toString()
  if (output?.length > MAX_STRING_LENGTH) {
    return output.substring(0, MAX_STRING_LENGTH) + '...'
  }
  return output ?? ''
}

export const jsonTruncate = (obj: unknown) => {
  return truncate(JSON.stringify(obj, null, 2))
}

export const timeTag = (datetime?: string) => {
  return datetime ? (
    <time dateTime={datetime} title={datetime}>
      {new Date(datetime).toUTCString()}
    </time>
  ) : (
    ''
  )
}

export const checkboxInputTag = (checked: boolean) => {
  return <input type="checkbox" checked={checked} disabled />
}
