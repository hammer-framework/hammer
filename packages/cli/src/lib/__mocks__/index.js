export const getPaths = () => MOCK_PATHS

export const MOCK_PATHS = {
  base: 'mock_basedir',
  generated: {
    base: '.mock_redwood',
    schema: '.mock_redwood/schema.graphql',
    types: {
      includes: '.mock_redwood/types/includes',
      mirror: '.mock_redwood/types/mirror',
    },
    prebuild: '.mock_redwood/prebuild',
  },
  scripts: 'mock_scripts_dir',
  api: {
    base: 'mock_api',
    dataMigrations: 'mock_dataMigrations',
    db: 'mock_db_dir',
    dbSchema: 'mock_db_schema_dir',
    functions: 'fmock_unctions_dir',
    graphql: 'mock_graphql_dir',
    lib: 'mock_lib_dir',
    generators: 'mock_generators_dir',
    config: 'mock_config_dir',
    services: 'mock_services_dir',
    directives: 'mock_directives_dir',
    src: 'mock_src_dir',
    dist: 'mock_api/dist',
    types: 'mock_api/types',
    models: 'mock_models_dir',
  },
  web: {
    routes: ['mock_mock/routes'],
    base: 'mock_web',
    pages: 'mock_pages_dir',
    components: 'mock_components_dir',
    layouts: 'mock_layouts_dir',
    src: 'mock_src_dir',
    generators: 'mock_generators_dir',
    app: 'mock_app_dir',
    index: 'mock_index_dir',
    config: 'mock_config_dir',
    webpack: 'mock_webpack_dir',
    postcss: 'mock_postcss_dir',
    storybookConfig: 'mock_storybookConfig_dir',
    storybookPreviewConfig: 'mock_storybookPreviewConfig_dir',
    storybookManagerConfig: 'mock_storybookManagerConfig_dir',
    dist: 'mock_web/dist',
    types: 'mock_web/types',
  },
}
