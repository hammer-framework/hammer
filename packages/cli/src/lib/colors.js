import chalk from 'chalk'

/**
 * To keep a consistent color/style palette between cli packages, such as
 * @redwood/cli and @redwood/create-redwood-app, please keep them compatible
 * with one and another. We'll might split up and refactor these into a
 * separate package when there is a strong motivation behind it.
 *
 * Current files:
 *
 * - packages/cli/src/lib/colors.js (this file)
 * - packages/create-redwood-app/src/create-redwood-app.js
 */
export default {
  error: chalk.bold.red,
  warning: chalk.keyword('orange'),
  success: chalk.greenBright,
  info: chalk.grey,

  bold: chalk.bold,
  underline: chalk.underline,

  green: chalk.green,

  header: chalk.bold.underline.hex('#e8e8e8'),
  cmd: chalk.hex('#808080'),
  redwood: chalk.hex('#ff845e'),
  love: chalk.redBright,
}
