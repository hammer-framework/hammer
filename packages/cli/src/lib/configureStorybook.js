import fs from 'fs-extra'

import { getPaths } from '.'

/**
 * Configure Storybook for the given template by creating a custom preview config
 */
export default function configureStorybook({ force }, newStorybookPreview) {
  const storybookPreviewConfigPath = getPaths().web.storybookPreviewConfig

  let storybookPreviewConfig
  /**
   *  Check if storybookPreviewConfigPath already exists.
   *  Merge both files if it does.
   *  By removing import react and export decorator from new config
   *  And put new config inside current config after last import
   */
  if (fs.existsSync(storybookPreviewConfigPath)) {
    if (force) {
      fs.unlinkSync(storybookPreviewConfigPath)
      storybookPreviewConfig = newStorybookPreview
    } else {
      const currentConfig = fs
        .readFileSync(storybookPreviewConfigPath)
        .toString()

      const newDecoratorsName = newStorybookPreview.match(
        /export const decorators = \[(.*?)\]/
      )[1]

      const currentDecoratorsName = currentConfig.match(
        /export const decorators = \[(.*?)\]/
      )[1]

      const decoratorsExport = `export const decorators = [${currentDecoratorsName}, ${newDecoratorsName}]`

      const insideNewStorybookConfigWithoutReactAndDecoration =
        newStorybookPreview
          .replace(/import \* as React from 'react'/, '')
          .replace(/export const decorators = .*/, '')

      const currentConfigWithoutDecoration = currentConfig.replace(
        /export const decorators = .*/,
        ''
      )

      const reversedCurrentConfig = currentConfigWithoutDecoration
        .split('\n')
        .reverse()

      const indexOfLastImport = reversedCurrentConfig.findIndex((value) =>
        /^import /.test(value)
      )
      reversedCurrentConfig.splice(
        indexOfLastImport,
        0,
        insideNewStorybookConfigWithoutReactAndDecoration
      )
      storybookPreviewConfig =
        reversedCurrentConfig.reverse().join(`\n`) +
        `\n` +
        currentConfig +
        `\n` +
        decoratorsExport
    }
  } else {
    storybookPreviewConfig = newStorybookPreview
  }

  fs.outputFileSync(storybookPreviewConfigPath, storybookPreviewConfig)
}
