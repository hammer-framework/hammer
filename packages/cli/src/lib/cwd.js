import { spawnSync } from 'child_process'
import path from 'path'

export const BASE_DIR = path.resolve(__dirname, '..', '..', '..', '..')
const CLI = path.join(BASE_DIR, 'packages', 'cli', 'dist', 'index.js')

export function rw(args, options) {
  const { status, stdout, stderr } = spawnSync('node', [CLI, ...args], {
    cwd: BASE_DIR,
    ...options,
  })

  return {
    status,
    stdout: stdout.toString().trim(),
    stderr: stderr.toString().trim(),
  }
}
