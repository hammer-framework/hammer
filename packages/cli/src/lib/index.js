import fs from 'fs'
import path from 'path'

import lodash from 'lodash/string'
import camelcase from 'camelcase'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'
import { paramCase } from 'param-case'
import { getDMMF } from '@prisma/sdk'
import { getPaths as getRedwoodPaths } from '@redwoodjs/internal'
import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'

import c from 'src/lib/colors'

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

/**
 * Returns the database schema for the given `name` database table parsed from
 * the schema.prisma of the target applicaiton. If no `name` is given then the
 * entire schema is returned.
 */
export const getSchema = async (name) => {
  const schemaPath = path.join(getPaths().api.db, 'schema.prisma')
  const metadata = await getDMMF({
    datamodel: readFile(schemaPath.toString()),
  })

  if (name) {
    const model = metadata.datamodel.models.find((model) => {
      return model.name === name
    })

    if (model) {
      return model
    } else {
      throw `No schema definition found for \`${name}\``
    }
  }

  return metadata.datamodel
}

/**
 * Returns variants of the passed `name` for usage in templates. If the given
 * name was "fooBar" then these would be:

 * pascalName: FooBar
 * singularPascalName: FooBar
 * pluralPascalName: FooBars
 * singularCamelName: fooBar
 * pluralCamelName: fooBars
 * singularParamName: foo-bar
 * pluralParamName: foo-bars
*/
const nameVariants = (name) => {
  const normalizedName = pascalcase(pluralize.singular(name))

  return {
    pascalName: pascalcase(name),
    camelName: camelcase(name),
    singularPascalName: normalizedName,
    pluralPascalName: pluralize(normalizedName),
    singularCamelName: camelcase(normalizedName),
    pluralCamelName: camelcase(pluralize(normalizedName)),
    singularParamName: paramCase(normalizedName),
    pluralParamName: paramCase(pluralize(normalizedName)),
  }
}

export const templateRoot = path.resolve(
  __dirname,
  '../commands/Generate/templates'
)

export const generateTemplate = (templateFilename, { name, ...rest }) => {
  const templatePath = path.join(templateRoot, templateFilename)
  const template = lodash.template(readFile(templatePath).toString())
  return template({
    name,
    ...nameVariants(name),
    ...rest,
  })
}

export const readFile = (target) => fs.readFileSync(target)

export const writeFile = async (
  target,
  contents,
  { overwriteExisting = false } = {}
) => {
  if (!overwriteExisting && fs.existsSync(target)) {
    throw new Error(`${target} already exists.`)
  }

  const filename = path.basename(target)
  const targetDir = target.replace(filename, '')
  fs.mkdirSync(targetDir, { recursive: true })
  fs.writeFileSync(target, contents)
}

export const bytes = (contents) => Buffer.byteLength(contents, 'utf8')

/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */
export const getPaths = () => {
  try {
    return getRedwoodPaths()
  } catch (e) {
    console.log(c.error(e.message))
    process.exit(0)
  }
}

/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const writeFilesTask = (files, options) => {
  const { base } = getPaths()
  return new Listr(
    Object.keys(files).map((file) => {
      const contents = files[file]
      return {
        title: `Writing \`./${path.relative(base, file)}\`...`,
        task: () => writeFile(file, contents, options),
      }
    })
  )
}

/**
 * Update the project's routes file.
 */
export const addRoutesToRouterTask = (routes) => {
  const redwoodPaths = getPaths()
  const routesContent = readFile(redwoodPaths.web.routes).toString()
  const newRoutesContent = routes.reduce((content, route) => {
    return content.replace(/(\s*)\<Router\>/, `$1<Router>$1  ${route}`)
  }, routesContent)
  writeFile(redwoodPaths.web.routes, newRoutesContent, {
    overwriteExisting: true,
  })
}

export const runCommandTask = async (commands, { verbose }) => {
  const tasks = new Listr(
    commands.map(({ title, cmd, args, opts = {} }) => ({
      title,
      task: async () => {
        return execa(cmd, args, {
          shell: true,
          cwd: `${getPaths().base}/api`,
          stdio: verbose ? 'inherit' : 'pipe',
          extendEnv: true,
          cleanup: true,
          ...opts,
        })
      },
    })),
    {
      renderer: verbose && VerboseRenderer,
      dateFormat: false,
    }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
  }
}

export * from './generateTempSchema'
