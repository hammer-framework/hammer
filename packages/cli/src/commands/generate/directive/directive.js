import path from 'path'

import camelcase from 'camelcase'
import execa from 'execa'
import Listr from 'listr'

import { getConfig } from '@redwoodjs/internal'

import { getPaths, writeFilesTask, transformTSToJS } from '../../../lib'
import c from '../../../lib/colors'
import {
  createYargsForComponentGeneration,
  templateForComponentFile,
} from '../helpers'

export const files = ({ name, typescript = false, tests }) => {
  if (tests === undefined) {
    tests = getConfig().generate.tests
  }

  const camelName = camelcase(name)

  const outputFilename = `${camelName}.directive.${typescript ? 'ts' : 'js'}`

  const directiveFile = templateForComponentFile({
    name,
    extension: typescript ? '.ts' : '.js',
    generator: 'directive',
    templatePath: 'directive.ts.template',
    outputPath: path.join(getPaths().api.directives, outputFilename),
    templateVars: { camelName },
  })

  const files = [directiveFile]

  if (tests) {
    const testOutputFilename = `${camelcase(name)}.test.${
      typescript ? 'ts' : 'js'
    }`

    const testFile = templateForComponentFile({
      name,
      extension: typescript ? '.test.ts' : '.test.js',
      generator: 'directive',
      templatePath: 'directive.test.ts.template',
      outputPath: path.join(getPaths().api.directives, testOutputFilename),
      templateVars: { camelName },
    })
    files.push(testFile)
  }

  // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }
  return files.reduce((acc, [outputPath, content]) => {
    const template = typescript ? content : transformTSToJS(outputPath, content)

    return {
      [outputPath]: template,
      ...acc,
    }
  }, {})
}

const positionalsObj = {
  name: {
    description: 'Name of your directive',
    type: 'string',
  },
}

export const { command, description, builder } =
  createYargsForComponentGeneration({
    componentName: 'directive',
    filesFn: files,
    positionalsObj,
  })

export const handler = async (args) => {
  const POST_RUN_INSTRUCTIONS = `Next steps...\n\n   ${c.warning(
    'After modifying your directive, you can add it to your SDLs e.g.:'
  )}
    ${c.info('// example todo.sdl.js')}
    ${c.info('# Option A: Add it to a field (transform or validation)')}
    type Todo {
      id: Int!
      body: String! ${c.green(`@${args.name}`)}
    }

    ${c.info('# Option B: Add it to query/mutation (validation only)')}
    type Query {
      todos: [Todo] ${c.green(`@${args.name}`)}
    }
`

  const tasks = new Listr(
    [
      {
        title: 'Generating directive file...',
        task: () => {
          return writeFilesTask(files(args), { overwriteExisting: args.force })
        },
      },
      {
        title: 'Generating redwood types...',
        task: () => {
          return execa('yarn rw-gen', [], {
            stdio: 'pipe',
            shell: true,
            cwd: getPaths().web.base,
          })
        },
      },
      {
        title: 'Next steps...',
        task: (_ctx, task) => {
          task.title = POST_RUN_INSTRUCTIONS
        },
      },
    ].filter(Boolean),
    { collapse: false }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
    process.exit(1)
  }
}
