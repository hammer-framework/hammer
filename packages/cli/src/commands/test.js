import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'

import { getPaths } from 'src/lib'
import c from 'src/lib/colors'

export const command = 'test [app..]'
export const desc = 'Run Jest tests for api and web.'
export const builder = {
  app: { choices: ['api', 'web'], default: ['api', 'web'] },
}

export const handler = async ({ app }) => {
  const { base: BASE_DIR } = getPaths()
  const execCommands = {
    api: {
      cwd: `${BASE_DIR}/api`,
      cmd: 'yarn jest',
      args: [
        '--passWithNoTests',
        '--config ../node_modules/@redwoodjs/core/config/jest.config.api.js',
      ],
    },
    web: {
      cwd: `${BASE_DIR}/web`,
      cmd: 'yarn jest',
      args: [
        '--passWithNoTests',
        '--config ../node_modules/@redwoodjs/core/config/jest.config.web.js',
      ],
    },
  }

  const tasks = new Listr(
    app.map((appName) => {
      const { cmd, args, cwd } = execCommands[appName]
      return {
        title: `Running '${appName}' jest tests`,
        task: () => {
          return execa(cmd, args, {
            stdio: 'inherit',
            shell: true,
            cwd,
          })
        },
      }
    }),
    {
      renderer: VerboseRenderer,
    }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
  }
}
