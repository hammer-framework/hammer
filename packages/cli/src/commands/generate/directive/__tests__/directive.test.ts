global.__dirname = __dirname
// Load shared mocks
import '../../../../lib/test'

import path from 'path'

import { files } from '../directive'

test('creates a JavaScript validator directive', () => {
  const output = files({
    name: 'require-admin', // checking camel casing too!
    typescript: false,
    tests: true,
    type: 'validator',
  })

  const expectedOutputPath = path.normalize(
    '/path/to/project/api/src/directives/requireAdmin/requireAdmin.js'
  )
  const expectedTestOutputPath = path.normalize(
    '/path/to/project/api/src/directives/requireAdmin/requireAdmin.test.js'
  )

  expect(Object.keys(output)).toContainEqual(expectedOutputPath)
  expect(Object.keys(output)).toContainEqual(expectedTestOutputPath)
  expect(output[expectedOutputPath]).toMatchSnapshot('js directive')
  expect(output[expectedTestOutputPath]).toMatchSnapshot('js directive test')
})

test('creates a TypeScript transformer directive', () => {
  const output = files({
    name: 'bazinga-foo_bar', // checking camel casing too!
    typescript: true,
    tests: true,
    type: 'transformer',
  })

  const expectedOutputPath = path.normalize(
    '/path/to/project/api/src/directives/bazingaFooBar/bazingaFooBar.ts'
  )
  const expectedTestOutputPath = path.normalize(
    '/path/to/project/api/src/directives/bazingaFooBar/bazingaFooBar.test.ts'
  )

  expect(Object.keys(output)).toContainEqual(expectedOutputPath)
  expect(Object.keys(output)).toContainEqual(expectedTestOutputPath)
  expect(output[expectedOutputPath]).toMatchSnapshot('ts directive')
  expect(output[expectedTestOutputPath]).toMatchSnapshot('ts directive test')
})
