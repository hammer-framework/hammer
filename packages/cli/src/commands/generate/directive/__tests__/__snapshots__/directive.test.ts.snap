// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a JavaScript directive 1`] = `
"export const schema = gql\`
  directive @requireAdmin on FIELD_DEFINITION
\`

export const requireAdmin = ({ context }, directiveNode) => {
  // Implement your directive here. This will be executed \\"onExecute\\"
  // More information on graphql lifecycle here: https://www.envelop.dev/docs/plugins/lifecycle

  // If you have auth setup you can access currentUser
  console.log(context.currentUser)

  // You can also modify your directive to take arguments
  // and use the getDirectiveArgument function from @redwoodjs/graphql-server
  // See documentation here: https://redwoodjs.com/docs/directives
}
"
`;

exports[`creates a TypeScript directive 1`] = `
"import type { DirectiveImplementationFunction } from '@redwoodjs/graphql-server'

export const schema = gql\`
  directive @bazingaFooBar on FIELD_DEFINITION
\`

export const bazingaFooBar: DirectiveImplementationFunction = (
  { context },
  directiveNode
) => {
  // Implement your directive here. This will be executed \\"onExecute\\"
  // More information on graphql lifecycle here: https://www.envelop.dev/docs/plugins/lifecycle

  // If you have auth setup you can access currentUser
  console.log(context.currentUser)

  // You can also modify your directive to take arguments
  // and use the getDirectiveArgument function from @redwoodjs/graphql-server
  // See documentation here: https://redwoodjs.com/docs/directives
}
"
`;
